devices/bme280/bme280.go

```diff
@@ -31,113 +34,209 @@ import (
+	F2       Filter = 1
+	F4       Filter = 2
+	F8       Filter = 3
+	F16      Filter = 4
 )
 
-// Dev is an handle to a bme280.
+// Dev is an handle to an initialized bme280.
 type Dev struct {
-	d     conn.Conn
-	isSPI bool
-	c     calibration
+	d         conn.Conn
+	isSPI     bool
+	opts      Opts
+	measDelay time.Duration
+	c         calibration
+
+	mu   sync.Mutex
+	stop chan struct{}
 }
 
 func (d *Dev) String() string {
 	return fmt.Sprintf("BME280{%s}", d.d)
 }
 
-// Sense returns measurements as °C, kPa and % of relative humidity.
+// Sense request a single one time measurement as °C, kPa and % of relative
+// humidity.
+//
+// The very first measurements may be of poor quality.
 func (d *Dev) Sense(env *devices.Environment) error {
-	// All registers must be read in a single pass, as noted at page 21, section
-	// 4.1.
-	// Pressure: 0xF7~0xF9
-	// Temperature: 0xFA~0xFC
-	// Humidity: 0xFD~0xFE
-	buf := [0xFF - 0xF7]byte{}
-	if err := d.readReg(0xF7, buf[:]); err != nil {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.stop != nil {
+		return errors.New("bme280: already sensing continuously")
+	}
+	err := d.writeCommands([]byte{
+		// ctrl_meas
+		0xF4, byte(d.opts.Temperature)<<5 | byte(d.opts.Pressure)<<2 | byte(forced),
+	})
+	if err != nil {
 		return err
 	}
-	// These values are 20 bits as per doc.
-	pRaw := int32(buf[0])<<12 | int32(buf[1])<<4 | int32(buf[2])>>4
-	tRaw := int32(buf[3])<<12 | int32(buf[4])<<4 | int32(buf[5])>>4
-	// This value is 16 bits as per doc.
-	hRaw := int32(buf[6])<<8 | int32(buf[7])
-
-	t, tFine := d.c.compensateTempInt(tRaw)
-	env.Temperature = devices.Celsius(t * 10)
-
-	p := d.c.compensatePressureInt64(pRaw, tFine)
-	env.Pressure = devices.KPascal((int32(p) + 127) / 256)
+	time.Sleep(d.measDelay)
+	for idle := false; !idle; {
+		if idle, err = d.isIdle(); err != nil {
+			return err
+		}
+	}
+	return d.sense(env)
+}
 
-	h := d.c.compensateHumidityInt(hRaw, tFine)
-	env.Humidity = devices.RelativeHumidity((int32(h)*100 + 511) / 1024)
-	return nil
+// SenseContinuous returns measurements as °C, kPa and % of relative humidity
+// on a continuous basis.
+//
+// Call Halt() to stop the sensing, which will close the channel.
+//
+// It's the responsibility of the caller to retrieve the values from the
+// channel as fast as possible.
+func (d *Dev) SenseContinuous(interval time.Duration, f Filter) (<-chan devices.Environment, error) {
+	d.mu.Lock()
+	defer d.mu.Unlock()
+	if d.stop != nil {
+		// Don't send the stop command to the device.
+		close(d.stop)
+		d.stop = nil
+	}
+	s := chooseStandby(interval - d.measDelay)
+	err := d.writeCommands([]byte{
+		// config
+		0xF5, byte(s)<<5 | byte(f)<<2,
+		// ctrl_meas
+		0xF4, byte(d.opts.Temperature)<<5 | byte(d.opts.Pressure)<<2 | byte(normal),
+	})
+	if err != nil {
+		return nil, err
+	}
+	sensing := make(chan devices.Environment)
+	d.stop = make(chan struct{})
+	go func() {
+		defer close(sensing)
+		d.senseContinuous(interval, sensing, d.stop)
```

Comment:
nit: I got confused not noticing that we were calling a different method and not making a recursive call. The recursive call didn't make a lot of sense (pun intended) and the args were different. A suggestions would be to call the unexported methods `sensingContinuously`.